### PR TITLE
Fix bug in RandomUniform Bias Initializer #417

### PR DESCRIPTION
**Pull request by WLeeeee**
The bug is mainly caused by calling random_uniform with minval == maxval.

rseed = 12345
model = keras.Sequential()
b = keras.initializers.RandomUniform(minval=-1.0, maxval=-1.0, seed=rseed)·
model.add(layers.Dense(16,
                            input_shape=(1,1000),
                            activation=keras.activations.elu,
                            use_bias=True,
                            bias_initializer=b,
                            )) 
print (model.layers[0].get_weights())

When minval = maxval, the tile language does the constant propagation and return a scalar.
However, the returned scalar is always zero no matter the value.
The reason is that we should return a tensor with broadcast zero when doing constant folding.

Below shows the aggressive optimization for the "mul" operation.
Here is an example of the compiled tile code with error:
The RevMul function only has one scalar operand.
 
DEBUG:plaidml:Doing a compilation of:
function (
 X_I_0[X_I_0_0, X_I_0_1]
 ) -> (
 X_T2,
 X_T3
 ) {
 X_T0 = 16;
 [[pid(PrngStep)]] X_T1 = prng_step(X_I_0, X_T0);
 [[pid(PrngState)]] X_T2 = prng_state(X_T1);
 [[pid(RevMul)]] X_T3 = -1.0;
}

DEBUG:plaidml:Compiling: function (
X_I_0[X_I_0_0, X_I_0_1]) -> (
X_T2,
X_T3
) {
X_T0 = 16;
[[pid(PrngStep)]] X_T1 = prng_step(X_I_0, X_T0);
[[pid(PrngState)]] X_T2 = prng_state(X_T1);
[[pid(Add)]] X_T3 = -1.0;
}

Here is the right tile code after the optimization:
X_I_0[X_I_0_0, X_I_0_1]
 ) -> (
X_T2,
X_T7
) {
X_T0 = 16;
[[pid(PrngStep)]] X_T1 = prng_step(X_I_0, X_T0);
[[pid(PrngState)]] X_T2 = prng_state(X_T1);
X_T3 = -1.0;
X_T4 = 0.0;
[[pid(PrngValue)]] X_T5 = prng_value(X_T1);
[[pid(RevMul)]] X_T6 = mul(X_T4, X_T5);
[[pid(Add)]] X_T7 = add(X_T3, X_T6);
}

### CODE REVIEW

File: tile/lang/compose.cc [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 #include <algorithm>
 #include <sstream>
+#include <stack>
```
Comment (dgkutnic) [inline tile/lang/compose.cc]: note (to Tim and just general FYI): this is a change that was requested by cpplint.

**Comment by WLeeeee** [comment]
A better way to fix this issue is to provide a general way to broadcast the constant to the target tensor with specified shape and dim. However, it requires non-trivial effort to do it. Thus, a simpler fix is to return the constant tensor in Keras backend when random_uniform is being called with minval == maxval.

File: tile/lang/compose.cc [hunk 2]
```diff
@@ -952,7 +956,11 @@ void FunctionApplication::SetDone() {
     const std::shared_ptr<Value> val = func_->in_bound().at(pi.name);
     if (pi.tag == Input::FIXED) {
       if (val->num_dims() != pi.dims.size()) {
-        throw std::runtime_error("Applying function, tensor with mismatching dimensionality: " + pi.name);
+        throw std::runtime_error(
+            "Applying function, tensor with mismatching dimensionality: " + pi.name +
+            "\n Check the \"Common Issues\" section on our troubleshooting documentation to see if this issue has "
```
Comment (dgkutnic) [inline tile/lang/compose.cc]: Are these changes specific to this PR?

Comment (tzerrell) [reply tile/lang/compose.cc]: Similarly, are there cases where we expect a user to hit one of these errors and information in the "Common Issues" section will help them?

Comment (dgkutnic) [reply tile/lang/compose.cc]: Our user was running into a `ValueError`, which may be caused by many reasons (of which this is just one).

These changes were actually from my `BatchDot` PR, which is why I'm confused that they appear here.

Comment (WLeeeee) [reply tile/lang/compose.cc]: I just copy and paste the up-to-date compose.cc from the website, which has those changes.
Actually, it does not belong to this fix.

Comment (flaub) [reply tile/lang/compose.cc]: I think what's needed here is a rebase or merge with master.

**Comment by WLeeeee** [comment]
Moved to:
fix random uniform problem #438
